components/places/sql/create_sync_triggers.sql

```diff
@@ -0,0 +1,171 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+-- This file defines triggers for the Sync connection.
+
+--- Pushes uploaded changes back to the local and remote trees. This is more
+--- or less equivalent to Desktop's `PlacesSyncUtils.bookmarks.pushChanges`.
+CREATE TEMP TRIGGER pushUploadedChanges
+AFTER UPDATE OF uploadedAt ON itemsToUpload WHEN NEW.uploadedAt > -1
+BEGIN
+  -- Reduce the change counter and update the sync status for uploaded items.
+  -- If the item was uploaded during the sync, its change counter will still
+  -- be > 0 for the next sync.
+  UPDATE moz_bookmarks SET
+      syncChangeCounter = max(syncChangeCounter - NEW.syncChangeCounter, 0),
+      syncStatus = 2 -- SyncStatus::Normal
+  WHERE guid = NEW.guid;
+
+  -- Remove uploaded tombstones.
+  DELETE FROM moz_bookmarks_deleted
+  WHERE guid = NEW.guid;
+
+  -- Write the uploaded item back to the synced bookmarks table, to match
+  -- what's on the server now.
+  REPLACE INTO moz_bookmarks_synced(guid, parentGuid, serverModified, needsMerge,
+                                    validity, isDeleted, kind, dateAdded, title,
+                                    placeId, keyword)
+  VALUES(NEW.guid, NEW.parentGuid, NEW.uploadedAt, 0,
+         1, -- SyncedBookmarkValidity::Valid
+         NEW.isDeleted, NEW.kind, NEW.dateAdded, NEW.title,
+         NEW.placeId, NEW.keyword);
+
+  INSERT INTO moz_bookmarks_synced_structure(guid, parentGuid, position)
+  SELECT guid, NEW.guid, position
+  FROM structureToUpload
+  WHERE parentId = NEW.id;
+END;
+
+-- Removes items that are deleted on one or both sides from local items,
+-- and inserts new tombstones for non-syncable items to delete remotely.
+CREATE TEMP TRIGGER removeLocalItems
+AFTER DELETE ON itemsToRemove
+BEGIN
+  -- Flag URL frecency for recalculation.
```

Comment:
Uh, do we actually recalculate negative frecencies anywhere? I also don't know that `frecency = -frecency` wont confuse the origin frecency update triggers badly...

Reply:
I added a `moz_places_stale_frecencies` table to hold Places with stale frecencies, but I haven't added logic to recalculate them yet. I wonder if we could, or should, do this on a background thread...or if we do this more like Desktop, where we backfill a fixed number (400) after every sync, or until maintenance runs, until we've recalculated everything.